### ISSUE TITLE
[web][CanvasKit] TextField with TextDirection.rtl has misplaced cursor

### ISSUE THREAD

**Issue by MarcinusX** [flutter/flutter#78550]
## Steps to Reproduce
**Only on Web**
Have a TextField with `textDirection` that mismatches keyboard text direction (e.g. English keyboard and `TextDireciton.rtl` ).
```
import 'package:flutter/material.dart';

void main() => runApp(RtlBrokenApp());

class RtlBrokenApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Padding(
          padding: const EdgeInsets.all(64),
          child: TextField(
            textDirection: TextDirection.rtl,
          ),
        ),
      ),
    );
  }
}
```

Reproducible on [DartPad](https://dartpad.dev/1e23cbb91c41d6e8f41df082295d3ccf?null_safety=true)
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/16286046/110289092-3815ea80-7fe9-11eb-89ad-a08013db4a48.gif)

**Expected results:** 
I expect the cursor to be always aligned to the right (like on ios/android).

**Actual results:** 
On the web, the cursor is in the wrong place compared to ios/android. It's ALMOST aligned to the right but not exactly (it's 1, 2 chars before it).

Reproducible on Fluter 2.1.0 and 2.0.2
Originally posted in https://github.com/flutter/flutter/issues/47745#issuecomment-792548177
CC @mdebbar

**Comment by nizar-khan** [comment]
Hello. First, I advise you to update to the latest version of the filters and try this simple code that will work with you
=====================================================================
class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Scaffold(
        body: Center(
          child: Container(
            width: 250,
            height: 100,
            child: Card(
              color: Colors.green,
              child: TextField(
                textAlign: TextAlign.left,
              ),
            ),
          ),
        ),
      ),
    );
  }
}

**Comment by ronytesler** [comment]
Any update? 
what's the best workaround for now?

**Comment by MarcinusX** [comment]
Fyi, it still occurs on master after #79015 was merged

**Comment by ronytesler** [comment]
@MarcinusX How can we make it P2 or P1?

**Comment by ronytesler** [comment]
It works OK on some of the browser's versions.

**Comment by TahaTesser** [comment]
Hey everyone!
I can confirm the issue still persists

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: Container(
          child: Directionality(
              textDirection: TextDirection.rtl, child: TextField()),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

``` 

</details>

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.4 at /Users/tahatesser/Code/flutter_stable
    • Framework revision b1395592de (6 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • IN2011 (mobile) • c9d8ee0c • android-arm64  • Android 11 (API 30)
    • macOS (desktop) • macos    • darwin-x64     • macOS 11.2.3 20D91
      darwin-x64
    • Chrome (web)    • chrome   • web-javascript • Google Chrome 89.0.4389.114

• No issues found!
```

```bash
[✓] Flutter (Channel master, 2.1.0-13.0.pre.499, on macOS 11.2.3 20D91
    darwin-x64, locale en-GB)
    • Flutter version 2.1.0-13.0.pre.499 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 3dbe7f23c5 (7 hours ago), 2021-04-07 14:04:02 +0800
    • Engine revision 92f4dfb048
    • Dart version 2.13.0 (build 2.13.0-208.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • IN2011 (mobile) • c9d8ee0c • android-arm64  • Android 11 (API 30)
    • macOS (desktop) • macos    • darwin-x64     • macOS 11.2.3 20D91
      darwin-x64
    • Chrome (web)    • chrome   • web-javascript • Google Chrome 89.0.4389.114

• No issues found!
```
	
</details>

**Comment by saharvx9** [comment]
Pls fix it :(

**Comment by nirbar89** [comment]
please do your best to help us!

**Comment by ronytesler** [comment]
@TahaTesser Is it a duplicate of [#32239](https://github.com/flutter/flutter/issues/32239)?

**Comment by TahaTesser** [comment]
@ronytesler 
No, not a duplicate

**Comment by TahaTesser** [comment]
<details> 
<summary>flutter doctor -v</summary> 

```console
[✓] Flutter (Channel master, 2.2.0-11.0.pre.247, on Microsoft Windows [Version 10.0.19042.928], locale en-US)
    • Flutter version 2.2.0-11.0.pre.247 at C:\Users\Taha\Code\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f55b2665e9 (7 hours ago), 2021-04-23 12:39:02 +0530
    • Engine revision 6fa0fb0059
    • Dart version 2.14.0 (build 2.14.0-18.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.9.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.9.31129.286
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.1.0)
    • Android Studio at C:\Users\Taha\Code\android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.3.3
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.55.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • IN2011 (mobile)   • c9d8ee0c • android-arm64  • Android 11 (API 30)
    • Windows (desktop) • windows  • windows-x64    • Microsoft Windows [Version 10.0.19042.928]
    • Chrome (web)      • chrome   • web-javascript • Google Chrome 90.0.4430.85
    • Edge (web)        • edge     • web-javascript • Microsoft Edge 89.0.774.77

! Doctor found issues in 1 category.
```
	
</details>

**Comment by ahmadod** [comment]
This issue is a deal-breaker to create Flutter RTL-based web apps.
Please kindly prioritize it

**Comment by ronytesler** [comment]
@ahmadod @tadaspetra I agree. Do you have an estimation for when it will be fixed so that we can at least know to better plan if to use Flutter Web or not?

**Comment by ronytesler** [comment]
@mdebbar Why is it closed? Was it fixed?

**Comment by MarcinusX** [comment]
I have tested it on `master` and it still occurs.  
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 2.3.0-17.0.pre.578, on macOS 11.4 20F71 darwin-x64, locale pl-PL)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.52.1)
[✓] Connected device (2 available)
```

**Comment by kiaxseventh** [comment]
I can confirm the issue still persists on stable and master channel,
 this issue also exist on web CanvasKit renderer and HTML renderer

**Comment by MarcinusX** [comment]
@mdebbar @TahaTesser Could you please verify and reopen to make sure we won't lose track of this one?

Thank you 🙏

**Comment by kiaxseventh** [comment]
This is a serious problem and needs to be followed up and fixed
@mdebbar @TahaTesser  please open the issue again
thanks

**Comment by hussamDana92** [comment]
i have the same issue

**Comment by TahaTesser** [comment]
Hi everyone, the issue indeed still reproduces on the latest `Channel master, 2.4.0-1.0.pre.143` on canvas kit, I don't see the issue in HTML.

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel master, 2.4.0-1.0.pre.143, on Microsoft Windows [Version 10.0.22000.65], locale en-US)
    • Flutter version 2.4.0-1.0.pre.143 at C:\Users\Taha\Code\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 311920b326 (9 hours ago), 2021-07-09 02:06:03 -0400
    • Engine revision dc06724099
    • Dart version 2.14.0 (build 2.14.0-291.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.2.0)
    • Android Studio at C:\Users\Taha\Code\android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] VS Code (version 1.58.0)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (5 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86     • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64     • Microsoft Windows [Version 10.0.22000.65]
    • Windows (UWP) (desktop) • winuwp        • windows-uwp-x64 •
    • Chrome (web)            • chrome        • web-javascript  • Google Chrome 91.0.4472.124
    • Edge (web)              • edge          • web-javascript  • Microsoft Edge 91.0.864.64

• No issues found!
```

</details>

**Comment by mdebbar** [comment]
Thanks for reporting. I'll take a look at this next week.

**Comment by darrenaustin** [comment]
This is also keeping a test skipped: 

https://github.com/flutter/flutter/blob/d3bdbed4985ca2ab174d65a6f556bd18ac03be5e/packages/flutter/test/widgets/editable_text_test.dart#L124

**Comment by MarcinusX** [comment]
@mdebbar @TahaTesser 
I can still reproduce in dartpad on flutter version 2.5.0.
Are there any plans to address this issue? It creates an awful experience for inputs in Arabic apps.

**Comment by mo-ah-dawood** [comment]
i also found it in version 2.5.0
when writing english in rtl layout

**Comment by saged87** [comment]
same problem !

**Comment by PhloxDev** [comment]
I have same problem in version 2.8.1

**Comment by razgino11** [comment]
Still having this issue on the latest flutter version

**Comment by saharvx9** [comment]
Still hapening in 2.10

**Comment by raghadalgwere** [comment]
Still happing in 2.10

**Comment by moesaeed** [comment]
Any updates?

**Comment by muhammadkamel** [comment]
Any updates?

Still facing the same issue

**Comment by mdebbar** [comment]
I'm unable to repro this issue on the HTML renderer. It only happens on CanvasKit.

cc @hterkelsen @yjbanov

**Comment by minusium** [comment]
Any workaround?

**Comment by MarcinusX** [comment]
Still reproducible in Flutter 3.

**Comment by bar4488** [comment]
> Any workaround?

for anyone still needing a solution, I've created a [simple package](https://pub.dev/packages/web_directional_text_field) who reimplements `TextField` and `TextFormField` with 2 new widgets, `WebTextField` and `WebTextFormField`. the implementation uses html `<input>` tag behind the scenes, so the behavior is identical to a browser web experience.

it does not have all the features of `TextField` or `TextFormField` yet, and is only a temporary solution, but I hope it will be helpful for anyone making RTL web apps 😄

**Comment by amir14a** [comment]
I faced same issue, please fix this

**Comment by a-v-ebrahimi** [comment]
Recently the issue appeared on Android too, previously it was just on the web.

**Comment by kiaxseventh** [comment]
this problem has existed for more than 6 months and the Flutter team has not done anything !!!

**Comment by kevmoo** [comment]
I'm diggin in here to see what can be done. Thanks for your patience!

**Comment by kevmoo** [comment]
For those following along here – please checkout master, if you can, and verify this addresses your concerns! Thanks!

**Comment by a-v-ebrahimi** [comment]
I checked master, on Android, the problem still exists.

Flutter 3.1.0-0.0.pre.1369 • channel master • https://github.com/flutter/flutter.git
Framework • revision 638a581046 (4 hours ago) • 2022-06-24 21:34:08 -0400
Engine • revision 4446e3edda
Tools • Dart 2.18.0 (build 2.18.0-224.0.dev) • DevTools 2.14.0

**Comment by kevmoo** [comment]
Web or native android?

**Comment by a-v-ebrahimi** [comment]
Native Android, I write a word, click on (A), the cursor jumps to one character before last (B):

![FV66akWXgAEw8K6](https://user-images.githubusercontent.com/179375/175759760-bb6210a0-32d1-43a8-9bb7-e454e00446a7.png)
,

**Comment by MarcinusX** [comment]
@kevmoo I can confirm the problem fixed on web on version:
`
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1372, on macOS 12.4 21F79 darwin-arm, locale en-PL)
`

Thank you! 🙏🙏🙏

However, I'm not sure if it's a direct follow-up of the issue, but when textfield is losing focus, the coursor stays in the same place. I'll create a separate issue for that as I don't think it's related to RTL.

**Comment by parisa7103** [comment]
> Native Android, I write a word, click on (A), the cursor jumps to one character before last (B):
> 
> ![FV66akWXgAEw8K6](https://user-images.githubusercontent.com/179375/175759760-bb6210a0-32d1-43a8-9bb7-e454e00446a7.png) ,

I have the same problem in RTL languages in flutter text fields on android devices. By dragging the cursor on the Farsi/Arabic texts, It is not possible to move it to the end of the text and it stuck in n-1 position. 

  `[√] Flutter (Channel master, 3.1.0-0.0.pre.1423, on Microsoft Windows [Version 10.0.19042.867], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[√] Chrome - develop for the web
[!] Visual Studio - develop for Windows (Visual Studio Enterprise 2022 17.2.4)
    X The current Visual Studio installation is incomplete. Please reinstall Visual Studio.
[√] Android Studio (version 2021.1)
[√] Android Studio (version 2021.3)
[√] VS Code, 64-bit edition (version 1.67.2)
[√] Connected device (4 available)
[√] HTTP Host Availability
`

**Comment by kevmoo** [comment]
Okay! So the fix we submitted here was for web. So we either need to reopen
this issue specific to native platforms or you should open a new issue
